src/adapter/src/catalog/builtin_table_updates.rs

```diff
@@ -1348,4 +1374,145 @@ impl CatalogState {
             .expect("privileges is 1 dimensional, and its length is used for the array length");
         row
     }
+
+    pub fn pack_statement_prepared_update(
+        &self,
+        record: &StatementPreparedRecord,
+    ) -> BuiltinTableUpdate {
+        let StatementPreparedRecord {
+            id,
+            session_id,
+            name,
+            sql,
+            prepared_at,
+        } = record;
+        let row = Row::pack_slice(&[
+            Datum::Uuid(*id),
+            Datum::Uuid(*session_id),
+            Datum::String(name.as_str()),
+            Datum::String(sql.as_str()),
+            Datum::TimestampTz(to_datetime(*prepared_at).try_into().expect("must fit")),
+        ]);
+        BuiltinTableUpdate {
+            id: self.resolve_builtin_table(&MZ_PREPARED_STATEMENT_HISTORY),
+            row,
+            diff: 1,
+        }
+    }
+
+    fn pack_statement_execution_inner(
+        &self,
+        record: &StatementBeganExecutionRecord,
+        packer: &mut RowPacker,
+    ) {
+        let StatementBeganExecutionRecord {
+            id,
+            prepared_statement_id,
+            sample_rate,
+            params,
+            began_at,
+        } = record;
+
+        packer.extend([
+            Datum::Uuid(*id),
+            Datum::Uuid(*prepared_statement_id),
+            Datum::Float64((*sample_rate).into()),
+        ]);
+        packer.push_list_with(|packer| {
+            for s in params {
+                packer.push::<Datum>(s.as_ref().map(String::as_str).into());
+            }
+        });
+        packer.push(Datum::TimestampTz(
+            to_datetime(*began_at).try_into().expect("Sane system time"),
+        ));
+    }
+
+    pub fn pack_statement_began_execution_update(
+        &self,
+        record: &StatementBeganExecutionRecord,
```

Comment:
IIRC the plan was to batch these locally first in order to amortize the writes. If this is still the case, why not accept a batch of `records` and a reusable `packer` in the new `pack_~` methods?

Reply:
We already have infrastructure for batching writes (added by Joe), so I am just using that.